File: frontend/src/services/research.ts
import { api } from "./api";

export interface SWOTItem {
    statement: string;
    evidence: string;
}

export interface SWOTAnalysis {
    strengths: SWOTItem[];
    weaknesses: SWOTItem[];
    opportunities: SWOTItem[];
    threats: SWOTItem[];
}

export interface NewsItem {
    title: string;
    source: string;
    date: string;
    url: string;
    summary?: string;
}

export interface MarketShare {
    name: string;
    share: number;
}

export interface SentimentScore {
    sentiment: string;
    score: number;
}

export interface SWOTScores {
    strengths: number;
    weaknesses: number;
    opportunities: number;
    threats: number;
}

export interface Metrics {
    competitor_market_share: MarketShare[] | null;
    swot_scores: SWOTScores | null;
    news_sentiment: SentimentScore[] | null;
}

export interface TimelineItem {
    year: string;
    title?: string;
    description?: string;
    category?: string;
    event?: string;
    source?: Source;
}

export interface Source {
    title: string;
    source_name: string;
    url: string;
    published_date?: string;
    content?: string;
}

export interface AccountPlanDetails {
    company_overview: string;
    products: string[];
    stakeholders: string[];
    challenges: string[];
    opportunities: string[];
    outreach_strategy: string;
}

export interface AccountPlanData {
    official_website: string | null;
    industry: string | null;
    company_type: string | null;
    executive_summary: string | null;
    executive_summary_sources: Source[];
    business_overview: string | null;
    products_services: string[];
    competitors: string[];
    competitors_sources: Source[];
    latest_news: NewsItem[];
    news_sources: Source[];
    swot: SWOTAnalysis | null;
    swot_sources: Source[];
    metrics: Metrics | null;
    timeline: TimelineItem[] | null;
    timeline_sources: Source[];
    account_plan: AccountPlanDetails | null;
    account_plan_sources: Source[];
}

export interface ResearchResponse {
    company: string;
    status: string;
    message: string | null;
    data?: AccountPlanData | null;
}

export async function researchCompany(
    company: string
): Promise<ResearchResponse> {
    const response = await api.post("/api/v1/research/", {
        company,
    });

    return response.data;
}

export interface ComparisonSection {
    company_a_val: string;
    company_b_val: string;
    comparison_synthesis: string;
    sources?: Source[];
}

export interface CompanyComparisonData {
    company_a: string;
    company_b: string;
    overview: ComparisonSection;
    market_position: ComparisonSection;
    swot: ComparisonSection;
    competitors: ComparisonSection;
    growth_potential: ComparisonSection;
    timeline: ComparisonSection;
    opportunities: ComparisonSection;
    risks: ComparisonSection;
}

export interface ComparisonResponse {
    status: string;
    message: string | null;
    data?: CompanyComparisonData | null;
    is_comparison?: boolean;
}

export async function compareCompanies(
    companyA: string,
    companyB: string
): Promise<ComparisonResponse> {
    const response = await api.post("/api/v1/research/compare", {
        company_a: companyA,
        company_b: companyB
    });
    if (response.data) {
        response.data.is_comparison = true;
    }
    return response.data;
}

File: backend/app/schemas/research.py
from pydantic import BaseModel
from typing import List, Dict, Any

class ResearchRequest(BaseModel):
    company: str

class SWOTItem(BaseModel):
    statement: str
    evidence: str

class SWOTAnalysis(BaseModel):
    strengths: List[SWOTItem]
    weaknesses: List[SWOTItem]
    opportunities: List[SWOTItem]
    threats: List[SWOTItem]

class NewsItem(BaseModel):
    title: str
    source: str
    date: str
    url: str
    summary: str | None = None

class MarketShare(BaseModel):
    name: str
    share: int

class SentimentScore(BaseModel):
    sentiment: str
    score: int

class SWOTScores(BaseModel):
    strengths: int
    weaknesses: int
    opportunities: int
    threats: int

class Metrics(BaseModel):
    competitor_market_share: List[MarketShare] | None
    swot_scores: SWOTScores | None
    news_sentiment: List[SentimentScore] | None

class Source(BaseModel):
    title: str
    source_name: str
    url: str
    published_date: str | None = None
    content: str | None = None

class TimelineItem(BaseModel):
    year: str
    title: str
    description: str
    category: str
    source: Source | None = None

class AccountPlanDetails(BaseModel):
    company_overview: str
    products: List[str]
    stakeholders: List[str]
    challenges: List[str]
    opportunities: List[str]
    outreach_strategy: str

class AccountPlanOutput(BaseModel):
    official_website: str | None
    industry: str | None
    company_type: str | None
    founder: str | None
    executive_summary: str | None
    executive_summary_sources: List[Source]
    business_overview: str | None
    products_services: List[str]
    competitors: List[str]
    competitors_sources: List[Source]
    latest_news: List[NewsItem]
    news_sources: List[Source]
    swot: SWOTAnalysis | None
    swot_sources: List[Source]
    metrics: Metrics | None
    timeline: List[TimelineItem]
    timeline_sources: List[Source]
    account_plan: AccountPlanDetails
    account_plan_sources: List[Source]

class ResearchResponse(BaseModel):
    company: str
    status: str
    message: str | None = None
    data: AccountPlanOutput | None = None

class ComparisonRequest(BaseModel):
    company_a: str
    company_b: str

class ComparisonSection(BaseModel):
    company_a_val: str
    company_b_val: str
    comparison_synthesis: str
    sources: List[Source] = []

class CompanyComparisonOutput(BaseModel):
    company_a: str
    company_b: str
    overview: ComparisonSection
    market_position: ComparisonSection
    swot: ComparisonSection
    competitors: ComparisonSection
    growth_potential: ComparisonSection
    timeline: ComparisonSection
    opportunities: ComparisonSection
    risks: ComparisonSection

class ComparisonResponse(BaseModel):
    status: str
    message: str | None = None
    data: CompanyComparisonOutput | None = None